clients/web/src/utilities/EventStream.js

```diff
@@ -14,103 +12,164 @@ var timestamp = null;
  * Listeners can bind to specific event types on the channel.
  */
 function EventStream(settings) {
-    var defaults = {
+    const defaults = {
         timeout: null,
-        streamPath: '/notification/stream',
-        _heartbeatTimeout: 5000 // in milliseconds
+        streamPath: '/notification/stream'
     };
 
+    // User-provided settings from initialization
     this.settings = _.extend(defaults, settings);
+    // Possible states are 'closed', 'stopped', 'started'
+    this._state = 'closed';
+    // Whether this is started (if so, holds the EventSource instance)
+    this._eventSource = null;
+
+    // Create context bindings only once, so they can be referenced when unbinding events
+    this._onVisibilityStateChange = _.bind(this._onVisibilityStateChange, this);
+    this._onMessage = _.bind(this._onMessage, this);
+    this._onError = _.bind(this._onError, this);
 
     return _.extend(this, Backbone.Events);
 }
 
-var prototype = EventStream.prototype;
-
-/*
- * This method is used to kill the event stream socket when the girder tab is not
- * visible using window.requestAnimationFrame. It creates a dead man switch to
- * close the frame after _heartbeatTimeout milliseconds (default of 5 seconds)
- * if requestAnimationFrame is not called in that time.
- */
-prototype._heartbeat = function () {
-    if (!this._stopHeartbeat) {
-        window.requestAnimationFrame(() => { this._heartbeat(); });
+EventStream.prototype._onVisibilityStateChange = function () {
+    if (document.visibilityState === 'visible' && this._state === 'stopped') {
+        this._start();
+    } else if (document.visibilityState === 'hidden' && this._state === 'started') {
+        this._stop();
     }
+};
 
-    if (this._eventSource) {
-        window.clearTimeout(this._closeTimeout);
-        this._closeTimeout = window.setTimeout(() => {
-            this.trigger('g:eventStream.stop');
-            this._eventSource.close();
-            this._eventSource = null;
-        }, this.settings._heartbeatTimeout);
-    } else {
-        this.open();
+EventStream.prototype._onMessage = function (e) {
+    let obj;
+    try {
+        obj = window.JSON.parse(e.data);
+    } catch (err) {
+        console.error('Invalid JSON from SSE stream: ' + e.data + ',' + err);
+        this.trigger('g:error', e);
+        return;
     }
+    EventStream.setLastTimestamp(obj._girderTime);
+    this.trigger('g:event.' + obj.type, obj);
```

Comment:
Can we simplify this control flow to:

```
try {
    const obj = window.JSON.parse(e.data);
    EventStream.setLastTimestamp(obj._girderTime);
    this.trigger('g:event.' + obj.type, obj);
} catch (err) {
    console.error('Invalid JSON from SSE stream: ' + e.data + ',' + err);
    this.trigger('g:error', e);
}

Reply:
I only want to catch JSON decoding errors, and Javascript doesn't have an `else` block for `try` statements (unlike Python).

Reply:
Fair enough, if we anticipate exceptions from the other two lines.